ngraph/test/backend/unary_test.hpp

```diff
@@ -0,0 +1,211 @@
+    void test_unary(std::shared_ptr<Function> f,
+                    const std::vector<value_type>& test_input,
+                    const std::vector<value_type>& test_expected,
+                    ArgShape ashape = {},
+                    ArgTolerance tol = {})
+    {
+        if (!ashape.initialzed)
+        {
+            ashape = ArgShape(Shape({test_input.size()}));
+        }
+        // the size specified in static_shape may be bigger or smaller than the
+        // number of data contained in test_input & test_expected, here we
+        // sample test data repetitively to generate actual test data of required size.
+        auto static_size = shape_size(ashape.static_shape);
+        std::vector<value_type> input;
+        std::vector<value_type> expected;
+        for (size_t i = 0; i < static_size; i++)
+        {
+            input.push_back(test_input[i % test_input.size()]);
+            expected.push_back(test_expected[i % test_expected.size()]);
+        }
+
+        auto test_case = test::TestCase<TestEngine>(f);
+        test_case.template add_input<value_type>(input);
+        test_case.template add_expected_output<value_type>(ashape.static_shape, expected);
+        if (tol.using_bits)
+        {
+            if (tol.tolerance_bits >= 0)
+                test_case.run(tol.tolerance_bits);
+            else
+                test_case.run();
+        }
+        else
+        {
+            test_case.run_with_tolerance_as_fp(tol.tolerance_fp);
+        }
+    }
+
+    // overloading support passing reference function instead of expected results
+    template <typename TestEngine,
+              element::Type_t element_type,
+              typename value_type = fundamental_type_for<element_type>>
+    void test_unary(std::shared_ptr<Function> f,
+                    const std::vector<value_type>& input,
+                    value_type (*func)(value_type),
+                    ArgShape ashape = {},
+                    ArgTolerance tol = {})
+    {
+        std::vector<value_type> expected;
+        for (value_type x : input)
+            expected.push_back(func(x));
+
+        test_unary<TestEngine, element_type>(f, input, expected, ashape, tol);
+    }
+
+    // overloading support passing creator instead of ngraph function
+    // Creator is any callable of following prototype:
+    //
+    //          std::shared_ptr<Function> (const ngraph::element::Type&, const PartialShape&)
+    //
+    // unary_func() is a helper function to generate a lambda as callable on simple unary ops with
+    // optional arguments captured in the that lambda.
+    //
+    using UnaryFuncCreator =
+        std::function<std::shared_ptr<Function>(const ngraph::element::Type&, const PartialShape&)>;
+
+    template <typename OpType, typename... Args>
+    typename std::enable_if<std::is_base_of<op::Op, OpType>::value, UnaryFuncCreator>::type
+        unary_func(Args&&... args)
```

Comment:
IMHO:
1) we should not mix template conditional instantiation with return type.
    https://godbolt.org/z/v7zKE6qE1
2) consider rename `unary_func` -> `create_unary_function`

Reply:
Agree